com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs

```diff
@@ -0,0 +1,161 @@
+using System.IO.Abstractions;
+using System.Text.RegularExpressions;
+using UnityEngine;
+using System.Collections.Generic;
+using System.IO;
+using MLAgents.Sensor;
+
+namespace MLAgents
+{
+    /// <summary>
+    /// Demonstration Recorder Component.
+    /// </summary>
+    [RequireComponent(typeof(BehaviorParameters))]
+    [AddComponentMenu("ML Agents/Demonstration Recorder", (int)MenuGroup.Default)]
+    public class DemonstrationRecorder : MonoBehaviour
+    {
+        [Tooltip("Whether or not to record demonstrations.")]
+        public bool record;
+
+        [Tooltip("Base demonstration file name. Will have numbers appended to make unique.")]
+        public string demonstrationName;
+
+        [Tooltip("Base directory to write the demo files. If null, will use {Application.dataPath}/Demonstrations.")]
+        public string demoDirectory;
+
+        DemonstrationStore m_DemoStore;
+        public const int MaxNameLength = 16;
+
+        const string k_ExtensionType = ".demo";
+        IFileSystem m_FileSystem;
+
+        void Start()
+        {
+            if (record)
+            {
+                InitializeDemoStore();
+            }
+        }
+
+        void Update()
+        {
+            if (record)
+            {
+                InitializeDemoStore();
+            }
+        }
+
+        /// <summary>
+        /// Creates demonstration store for use in recording.
+        /// </summary>
+        public void InitializeDemoStore(IFileSystem fileSystem = null)
+        {
+            if (m_DemoStore != null)
+            {
+                return;
+            }
+
+            m_FileSystem = fileSystem ?? new FileSystem();
+            var behaviorParams = GetComponent<BehaviorParameters>();
+            if (string.IsNullOrEmpty(demonstrationName))
+            {
+                demonstrationName = behaviorParams.behaviorName;
+            }
+            if (string.IsNullOrEmpty(demoDirectory))
+            {
+                demoDirectory = Path.Combine(Application.dataPath, "Demonstrations");
+            }
+
+            demonstrationName = SanitizeName(demonstrationName, MaxNameLength);
+            CreateDirectory();
+            var stream = CreateDemonstrationFile(demonstrationName);
+            m_DemoStore = new DemonstrationStore(stream);
+
+            m_DemoStore.Initialize(
+                demonstrationName,
+                behaviorParams.brainParameters,
+                behaviorParams.fullyQualifiedBehaviorName
+            );
+            var recordingAgent = GetComponent<Agent>();
+            Monitor.Log("Recording Demonstration of Agent: ", recordingAgent?.name);
```

Comment:
if the Agent is null, this will still display "Recording Demonstration of Agent :". If Agent is never null (although we never really know) then go for recordingAgent ~?~ .name

Reply:
Should probably just remove it. This is the only place the Agent is used. Previously there was `RequireComponent(Agent)` but that seems unnecessary.